Avoid copying for return values from swift

nmahendru opened the pull request:
#### 20dee432499d1fd7fa1abcaf3f3a408989a68a95
<pre>
Avoid copying for return values from swift
<a href="https://bugs.webkit.org/show_bug.cgi?id=275619">https://bugs.webkit.org/show_bug.cgi?id=275619</a>
<a href="https://rdar.apple.com/129772363">rdar://129772363</a>

Reviewed by David Kilzer.

In the existing Swift-Cpp interop interface, when C++ tries to extract values out of swift structs, it generates a Copy as swift structs are value types.
Even if we declare the same structure in C++ and construct it on swift side with std::optional in it, and then get the value on C++ side, it still generates a copy. That&apos;s because
`Cpp.makeOptional` that is being removed here cannot take an rvalue reference(no support in swift).

With the current interop features, the best solution for now is to remove optionals and return an empty vector in error scenarios.
This should be a reasonable mitigation as the hot path &quot;should&quot; be success in which case the vector returned actually holds a useful value.
In errors, yes, an empty vector will be returned(adding std::optional was a solution for that problem) but that seems to add extra copies which we
would like to avoid if possible.

* Source/WebCore/PAL/pal/PALSwift.h:
(Cpp::makeOptional): Deleted.
* Source/WebCore/PAL/pal/PALSwift/CryptoKitShim.swift:
(AesKw.wrap(_:using:)):
(AesKw.unwrap(_:using:)):
(ECImportReturnValue.errorCode):
(ECKey.importX963Pub(_:curve:)):
(ECKey.exportX963Pub):
(ECKey.importCompressedPub(_:curve:)):
(ECKey.importX963Private(_:curve:)):
(ECKey.exportX963Private):
(ECKey.sign(_:hashFunction:)):
(ECKey.deriveBits(_:)):
(AesGcmReturnValue.cipherText): Deleted.
(AesGcmReturnValue.errorCode): Deleted.
(AesKwReturnValue.errorCode): Deleted.
(AesKwReturnValue.result): Deleted.
(ECReturnValue.errorCode): Deleted.
(ECReturnValue.signature): Deleted.
(ECReturnValue.keyBytes): Deleted.
(ECReturnValue.key): Deleted.
(EdReturnValue.errorCode): Deleted.
(EdReturnValue.signature): Deleted.
(EdReturnValue.keyBytes): Deleted.
(HKDFReturnValue.errorCode): Deleted.
(HKDFReturnValue.key): Deleted.
* Source/WebCore/crypto/cocoa/CryptoAlgorithmAESGCMMac.cpp:
(WebCore::encryptCryptoKitAESGCM):
* Source/WebCore/crypto/cocoa/CryptoAlgorithmAESKWMac.cpp:
(WebCore::wrapKeyAESKWCryptoKit):
(WebCore::unwrapKeyAESKWCryptoKit):
* Source/WebCore/crypto/cocoa/CryptoAlgorithmECDHMac.cpp:
(WebCore::platformDeriveBitsCC):
(WebCore::platformDeriveBitsCryptoKit):
* Source/WebCore/crypto/cocoa/CryptoAlgorithmECDSAMac.cpp:
(WebCore::signECDSACryptoKit):
(WebCore::verifyECDSACryptoKit):
* Source/WebCore/crypto/cocoa/CryptoAlgorithmEd25519Cocoa.cpp:
(WebCore::signEd25519CryptoKit):
(WebCore::verifyEd25519CryptoKit):
* Source/WebCore/crypto/cocoa/CryptoAlgorithmHKDFMac.cpp:
(WebCore::platformDeriveBitsCryptoKit):
* Source/WebCore/crypto/cocoa/CryptoAlgorithmX25519Cocoa.cpp:
(WebCore::deriveBitsCryptoKit):
* Source/WebCore/crypto/cocoa/CryptoKeyECMac.cpp:
(WebCore::CryptoKeyEC::platformImportRaw):
(WebCore::CryptoKeyEC::platformExportRaw const):
(WebCore::CryptoKeyEC::platformImportJWKPrivate):
(WebCore::CryptoKeyEC::platformAddFieldElements const):
(WebCore::CryptoKeyEC::platformImportSpki):
(WebCore::CryptoKeyEC::platformExportSpki const):
(WebCore::CryptoKeyEC::platformImportPkcs8):
(WebCore::CryptoKeyEC::platformExportPkcs8 const):

Canonical link: <a href="https://commits.webkit.org/280218@main">https://commits.webkit.org/280218@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a4a46a914fbf6b76deb7c668a60532cd75441aa7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/55981 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/35305 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/8449 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/58980 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/6413 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/58107 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/42926 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/6609 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/45085 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/4436 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/58010 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/33200 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/48273 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/26223 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/29982 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/5601 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/4556 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/51968 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/5872 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/60568 "Built successfully") | 
| | [⏳ 🛠 vision ](https://ews-build.webkit.org/#/builders/visionOS-1-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/5993 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/52512 "Passed tests") | 
| | [⏳ 🛠 vision-sim ](https://ews-build.webkit.org/#/builders/visionOS-1-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/48342 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/52044 "run-api-tests-without-change (failure)") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-1-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8291 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/31134 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/32218 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/33300 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/31966 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

ddkilzer commented on Source/WebCore/PAL/pal/PALSwift.h:
```diff
@@ -34,12 +34,29 @@ using VectorUInt8 = WTF::Vector<uint8_t>;
+    UnsupportedAlgorithm,
+};
+struct CryptoOperationReturnValue {
+    ErrorCodes errorCode = ErrorCodes::DefaultValue;
```
Nit:  Typically we initialize using `{ }` braces.

```suggestion
    ErrorCodes errorCode { ErrorCodes::DefaultValue };
```

ddkilzer reviewed (approved):
r=me.  (My eyes glazed over a bit when reviewing—especially the Swift code—so you may want to get another review.  Not required, though.)

achristensen07 commented on Source/WebCore/crypto/cocoa/CryptoAlgorithmECDHMac.cpp:
```diff
@@ -49,7 +49,7 @@ static std::optional<Vector<uint8_t>> platformDeriveBitsCC(const CryptoKeyEC& ba
     if (!CCECCryptorComputeSharedSecret(baseKey.platformKey().get(), publicKey.platformKey().get(), derivedKey.data(), &size))
 #endif
-        result = WTFMove(derivedKey);
+        result = std::make_optional(WTFMove(derivedKey));
```
std::make_optional isn't needed in many cases.

nmahendru committed:
Avoid copying for return values from swift
https://bugs.webkit.org/show_bug.cgi?id=275619
rdar://129772363

Reviewed by David Kilzer.

In the existing Swift-Cpp interop interface, when C++ tries to extract values out of swift structs, it generates a Copy as swift structs are value types.
Even if we declare the same structure in C++ and construct it on swift side with std::optional in it, and then get the value on C++ side, it still generates a copy. That's because
`Cpp.makeOptional` that is being removed here cannot take an rvalue reference(no support in swift).

With the current interop features, the best solution for now is to remove optionals and return an empty vector in error scenarios.
This should be a reasonable mitigation as the hot path "should" be success in which case the vector returned actually holds a useful value.
In errors, yes, an empty vector will be returned(adding std::optional was a solution for that problem) but that seems to add extra copies which we
would like to avoid if possible.

* Source/WebCore/PAL/pal/PALSwift.h:
(Cpp::makeOptional): Deleted.
* Source/WebCore/PAL/pal/PALSwift/CryptoKitShim.swift:
(AesKw.wrap(_:using:)):
(AesKw.unwrap(_:using:)):
(ECImportReturnValue.errorCode):
(ECKey.importX963Pub(_:curve:)):
(ECKey.exportX963Pub):
(ECKey.importCompressedPub(_:curve:)):
(ECKey.importX963Private(_:curve:)):
(ECKey.exportX963Private):
(ECKey.sign(_:hashFunction:)):
(ECKey.deriveBits(_:)):
(AesGcmReturnValue.cipherText): Deleted.
(AesGcmReturnValue.errorCode): Deleted.
(AesKwReturnValue.errorCode): Deleted.
(AesKwReturnValue.result): Deleted.
(ECReturnValue.errorCode): Deleted.
(ECReturnValue.signature): Deleted.
(ECReturnValue.keyBytes): Deleted.
(ECReturnValue.key): Deleted.
(EdReturnValue.errorCode): Deleted.
(EdReturnValue.signature): Deleted.
(EdReturnValue.keyBytes): Deleted.
(HKDFReturnValue.errorCode): Deleted.
(HKDFReturnValue.key): Deleted.
* Source/WebCore/crypto/cocoa/CryptoAlgorithmAESGCMMac.cpp:
(WebCore::encryptCryptoKitAESGCM):
* Source/WebCore/crypto/cocoa/CryptoAlgorithmAESKWMac.cpp:
(WebCore::wrapKeyAESKWCryptoKit):
(WebCore::unwrapKeyAESKWCryptoKit):
* Source/WebCore/crypto/cocoa/CryptoAlgorithmECDHMac.cpp:
(WebCore::platformDeriveBitsCC):
(WebCore::platformDeriveBitsCryptoKit):
* Source/WebCore/crypto/cocoa/CryptoAlgorithmECDSAMac.cpp:
(WebCore::signECDSACryptoKit):
(WebCore::verifyECDSACryptoKit):
* Source/WebCore/crypto/cocoa/CryptoAlgorithmEd25519Cocoa.cpp:
(WebCore::signEd25519CryptoKit):
(WebCore::verifyEd25519CryptoKit):
* Source/WebCore/crypto/cocoa/CryptoAlgorithmHKDFMac.cpp:
(WebCore::platformDeriveBitsCryptoKit):
* Source/WebCore/crypto/cocoa/CryptoAlgorithmX25519Cocoa.cpp:
(WebCore::deriveBitsCryptoKit):
* Source/WebCore/crypto/cocoa/CryptoKeyECMac.cpp:
(WebCore::CryptoKeyEC::platformImportRaw):
(WebCore::CryptoKeyEC::platformExportRaw const):
(WebCore::CryptoKeyEC::platformImportJWKPrivate):
(WebCore::CryptoKeyEC::platformAddFieldElements const):
(WebCore::CryptoKeyEC::platformImportSpki):
(WebCore::CryptoKeyEC::platformExportSpki const):
(WebCore::CryptoKeyEC::platformImportPkcs8):
(WebCore::CryptoKeyEC::platformExportPkcs8 const):

Canonical link: https://commits.webkit.org/280218@main
